mqtt/mqtt-broker/src/server.rs

```diff
@@ -56,36 +68,84 @@ where
 
                 // shutdown the incoming loop
                 info!("shutting down accept loop...");
-                itx.send(()).unwrap();
+
+                debug!("sending stop signal for every protocol head");
+                send_shutdown(shutdown_handles);
+
                 match tasks.await {
-                    Either::Right((_, broker_task)) => {
+                    Either::Right(((result, _index, unfinished_incoming_tasks), broker_task)) => {
+                        // wait until the rest of incoming_tasks finished
+                        let mut results = vec![result];
+                        results.extend(future::join_all(unfinished_incoming_tasks).await);
+
+                        for e in results.into_iter().filter_map(Result::err) {
+                            warn!(message = "failed to shutdown protocol head", error=%e);
+                        }
+
                         debug!("sending Shutdown message to broker");
                         handle.send(Message::System(SystemEvent::Shutdown)).await?;
                         broker_task.await.context(ErrorKind::TaskJoin)?
                     }
-                    Either::Left((broker_state, incoming_task)) => {
+                    Either::Left((broker_state, incoming_tasks)) => {
                         warn!("broker exited before accept loop");
-                        incoming_task.await?;
+
+                        // wait until either of incoming_tasks finished
+                        let (result, _index, unfinished_incoming_tasks) = incoming_tasks.await;
+
+                        // wait until the rest of incoming_tasks finished
+                        let mut results = vec![result];
+                        results.extend(future::join_all(unfinished_incoming_tasks).await);
+
+                        for e in results.into_iter().filter_map(Result::err) {
+                            warn!(message = "failed to shutdown protocol head", error=%e);
+                        }
+
                         broker_state.context(ErrorKind::TaskJoin)?
                     }
                 }
             }
-            Either::Right((either, _shutdown)) => match either {
-                Either::Right((Ok(_incoming_task), broker_task)) => {
-                    debug!("sending Shutdown message to broker");
-                    handle.send(Message::System(SystemEvent::Shutdown)).await?;
-                    broker_task.await.context(ErrorKind::TaskJoin)?
-                }
-                Either::Right((Err(e), broker_task)) => {
-                    error!(message = "an error occurred in the accept loop", error=%e);
+            Either::Right((either, _)) => match either {
```

Comment:
it's really hard to follow all these nested selects... some inline comments would be nice :)